components/license-control-service/pkg/server/server.go

```diff
@@ -274,90 +142,106 @@ func (s *LicenseControlServer) Update(ctx context.Context, req *lc.UpdateRequest
 		return nil, status.Errorf(codes.InvalidArgument, err.Error())
 	}
 
+	licensedPeriod := licensedPeriod(license)
+	logctx := log.WithFields(log.Fields{
+		"license_id": license.Id,
+		"expired_at": licensedPeriod.end,
+	})
+
 	span.LogFields(
 		spanlog.String("submitted_license", licData),
 		spanlog.Object("loaded_license", license),
-		spanlog.Object("loaded_policy", policy),
 		spanlog.Object("licensed_period", licensedPeriod),
 	)
 
 	if !req.Force && ptypes.TimestampNow().GetSeconds() > licensedPeriod.end.GetSeconds() {
 		msg := fmt.Sprintf("Rejecting license ID %v, expired at %v; set force=true to override",
 			license.Id, licensedPeriod.end)
-		log.WithFields(
-			log.Fields{
-				"license_id": license.Id,
-				"expired_at": licensedPeriod.end,
-			},
-		).Warn(msg)
-
+		logctx.Warn(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
-	} else {
-		msg := fmt.Sprintf("Applying license ID %v, expired at %v", license.Id, licensedPeriod.end)
-		log.WithFields(
-			log.Fields{
-				"license_id": license.Id,
-				"expired_at": licensedPeriod.end,
-			},
-		).Warn(msg)
 	}
 
-	// If a license is already loaded, and its ID matches the submitted license
-	if s.license != nil && license.Id == s.license.Id {
-		log.WithFields(
-			log.Fields{"license_id": s.license.Id},
-		).Warn("Submitted license already configured")
-
-		// Unless force is true...
-		if !req.Force {
-			return &lc.UpdateResponse{
-					Updated:   false,
-					Message:   fmt.Sprintf("License ID %s already set", license.Id),
-					Duplicate: true,
-				},
-				nil
-		}
-
-		log.WithFields(
-			log.Fields{"license_id": s.license.Id},
-		).Warn("License reconfigure forced")
+	logctx.Info("Applying license")
+	err = s.persistLicenseToken(ctx, licData)
+	switch err.(type) {
+	case nil:
+		return &lc.UpdateResponse{
+			Updated: true,
+			Message: fmt.Sprintf("Policy configured from license ID %s", license.Id),
+		}, nil
+	case *storage.RetriableBackendError:
+		return nil, status.Error(codes.Aborted, err.Error())
+	default:
+		logctx.WithError(err).Error("Unable to persist license")
+		return nil, status.Errorf(codes.Internal, "unable to persist license: %s", err.Error())
 	}
+}
 
-	// Persist the license to disk
-	if err := s.persistLicenseToken(ctx, licData); err != nil {
-		log.WithFields(
-			log.Fields{
-				"license_id": license.Id,
-				"error":      err,
-			},
-		).Error("Unable to persist license")
+//Telemetry endpoint to return telemetry configuration
+func (s *LicenseControlServer) Telemetry(ctx context.Context, req *lc.TelemetryRequest) (*lc.TelemetryResponse, error) {
+	return &lc.TelemetryResponse{TelemetryEnabled: s.TelemetryEnabled, TelemetryUrl: s.TelemetryURL}, nil
```

Comment:
is this on disk state that we will need to care about?

Reply:
This is something that needs discussion.  The *current* method of setting this is via the config:

https://automate.chef.io/docs/telemetry/#opting-out-of-telemetry

however, there is a legacy on-disk state for telemetry that the configuration looks for at server startup.